**Author Name:** *Ahsanullah Ghani*

**Email:** *ahsanullahghani000@gmail.com*

 **The data was Collected from the following:** [Link](https://www.kaggle.com/datasets/harlfoxem/housesalesprediction)

<style>
  .dataset-info {
    font-family: 'Segoe UI', Arial, sans-serif;
    background: #f9fafb;
    border: 1px solid #e5e7eb;
    border-radius: 12px;
    padding: 24px 28px;
    max-width: 800px;
    margin: 20px auto;
    color: #1f2937;
    line-height: 1.7;
  }
  .dataset-info h2 {
    font-weight: bold;
    font-size: 1.6rem;
    color: #2563eb;
    border-bottom: 3px solid #2563eb;
    padding-bottom: 8px;
    margin-bottom: 16px;
  }
  .dataset-info h3 {
    font-weight: bold;
    font-size: 1.2rem;
    color: #0ea5e9;
    margin-top: 24px;
  }
  .dataset-info table {
    width: 100%;
    border-collapse: collapse;
    margin-top: 12px;
    font-size: 0.95rem;
  }
  .dataset-info th {
    background: #2563eb;
    color: white;
    font-weight: bold;
    text-align: left;
    padding: 10px 12px;
  }
  .dataset-info td {
    padding: 8px 12px;
    border-bottom: 1px solid #e5e7eb;
  }
  .dataset-info tr:nth-child(even) {
    background: #f3f4f6;
  }
  .badge {
    display: inline-block;
    background: linear-gradient(135deg, #2563eb, #0ea5e9);
    color: white;
    font-weight: bold;
    font-size: 0.8rem;
    padding: 4px 12px;
    border-radius: 20px;
    margin: 4px 4px 0 0;
  }
</style>

<div class="dataset-info">
  <h2>📊 About the Dataset</h2>
  <p>This project uses the <b>King County House Sales dataset</b> (<code>kc_house_data.csv</code>), containing real estate transaction records for homes sold in King County, Washington, USA (Seattle area) between 2014 and 2015.</p>

  <h3>Dataset Summary</h3>
  <ul>
    <li><b>Total records:</b> 21,613 house sales</li>
    <li><b>Total features:</b> 21 columns</li>
    <li><b>Missing values:</b> None (fully complete)</li>
    <li><b>Target variable:</b> <code>price</code> — the sale price of the house (USD)</li>
  </ul>

  <h3>Feature Description</h3>
  <table>
    <tr><th>Feature</th><th>Description</th></tr>
    <tr><td>id</td><td>Unique identifier for each house</td></tr>
    <tr><td>date</td><td>Date the house was sold</td></tr>
    <tr><td>price</td><td>Sale price of the house (target variable)</td></tr>
    <tr><td>bedrooms</td><td>Number of bedrooms</td></tr>
    <tr><td>bathrooms</td><td>Number of bathrooms</td></tr>
    <tr><td>sqft_living</td><td>Living area in square feet</td></tr>
    <tr><td>sqft_lot</td><td>Lot (plot) size in square feet</td></tr>
    <tr><td>floors</td><td>Number of floors</td></tr>
    <tr><td>waterfront</td><td>1 if waterfront property, else 0</td></tr>
    <tr><td>view</td><td>Quality of view (rated 0–4)</td></tr>
    <tr><td>condition</td><td>Overall condition (rated 1–5)</td></tr>
    <tr><td>grade</td><td>Construction/design grade (rated 1–13)</td></tr>
    <tr><td>sqft_above</td><td>Square footage above ground</td></tr>
    <tr><td>sqft_basement</td><td>Square footage of basement</td></tr>
    <tr><td>yr_built</td><td>Year house was built</td></tr>
    <tr><td>yr_renovated</td><td>Year of last renovation (0 = never)</td></tr>
    <tr><td>zipcode</td><td>ZIP code of property</td></tr>
    <tr><td>lat</td><td>Latitude coordinate</td></tr>
    <tr><td>long</td><td>Longitude coordinate</td></tr>
    <tr><td>sqft_living15</td><td>Avg living area of 15 nearest neighbors</td></tr>
    <tr><td>sqft_lot15</td><td>Avg lot size of 15 nearest neighbors</td></tr>
  </table>

  <h3>Purpose</h3>
  <p>This dataset is commonly used for <b>regression analysis and predictive modeling</b> — predicting house <code>price</code> from physical features, condition, and location. It's also well suited for EDA, feature engineering, and geospatial visualization.</p>

  <div>
    <span class="badge">Regression</span>
    <span class="badge">EDA</span>
    <span class="badge">Feature Engineering</span>
    <span class="badge">Geo-Visualization</span>
  </div>
</div>

<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **TASK**

</div>

<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **1. BASIC MATH APPLICATION**

</div>

In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats

In [2]:
df = pd.read_csv(r"C:\Users\AHSAN GHANI\Documents\6Month Intership program\Frist Month\pandas_tip&trick\Data set\House Prices Dataset\kc_house_data.csv")

In [3]:
df.head()

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7129300520,20141013T000000,221900.0,3,1.00,1180,5650,1.0,0,0,...,7,1180,0,1955,0,98178,47.5112,-122.257,1340,5650
1,6414100192,20141209T000000,538000.0,3,2.25,2570,7242,2.0,0,0,...,7,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639
2,5631500400,20150225T000000,180000.0,2,1.00,770,10000,1.0,0,0,...,6,770,0,1933,0,98028,47.7379,-122.233,2720,8062
3,2487200875,20141209T000000,604000.0,4,3.00,1960,5000,1.0,0,0,...,7,1050,910,1965,0,98136,47.5208,-122.393,1360,5000
4,1954400510,20150218T000000,510000.0,3,2.00,1680,8080,1.0,0,0,...,8,1680,0,1987,0,98074,47.6168,-122.045,1800,7503


In [5]:
# 1) PRICE CHANGE % - how much price changed compared to the previous row
df['price_pct_change'] = df['price'].pct_change() * 100

# 2) RATIO - price for each square foot
df['price_per_sqft'] = df['price'] / df['sqft_living']

# 3) RATIO - average price for each bedroom
df['price_per_bedroom'] = df['price'] / df['bedrooms']

print(df[['price', 'sqft_living', 'price_pct_change', 'price_per_sqft', 'price_per_bedroom']].head(10))
print("\nAverage price change %:", df['price_pct_change'].mean())
print("Average price per sqft:", df['price_per_sqft'].mean())

       price  sqft_living  price_pct_change  price_per_sqft  price_per_bedroom
0   221900.0         1180               NaN      188.050847       73966.666667
1   538000.0         2570        142.451555      209.338521      179333.333333
2   180000.0          770        -66.542751      233.766234       90000.000000
3   604000.0         1960        235.555556      308.163265      151000.000000
4   510000.0         1680        -15.562914      303.571429      170000.000000
5  1225000.0         5420        140.196078      226.014760      306250.000000
6   257500.0         1715        -78.979592      150.145773       85833.333333
7   291850.0         1060         13.339806      275.330189       97283.333333
8   229500.0         1780        -21.363714      128.932584       76500.000000
9   323000.0         1890         40.740741      170.899471      107666.666667

Average price change %: 31.64086381567157
Average price per sqft: 264.1565936945145


 **Work with fractions and proportions from data**

In [6]:
# how many houses have more than 3 bedrooms - getting the fraction
fraction_more_than_3bed = (df['bedrooms'] > 3).sum() / len(df)

# how many houses have 2 or more bathrooms
fraction_2plus_bath = (df['bathrooms'] >= 2).sum() / len(df)

# bathroom to bedroom ratio for each house - gives an idea of house size
df['bath_bed_ratio'] = df['bathrooms'] / df['bedrooms']

# each house's sqft as a share of the total sqft (all shares add up to 100%)
df['sqft_share'] = df['sqft_living'] / df['sqft_living'].sum()

print(f"Houses with more than 3 bedrooms: {fraction_more_than_3bed:.2%}")
print(f"Houses with 2+ bathrooms: {fraction_2plus_bath:.2%}")
print(df[['bedrooms', 'bathrooms', 'bath_bed_ratio', 'sqft_share']].head(10))

Houses with more than 3 bedrooms: 40.79%
Houses with 2+ bathrooms: 60.94%
   bedrooms  bathrooms  bath_bed_ratio  sqft_share
0         3       1.00        0.333333    0.000026
1         3       2.25        0.750000    0.000057
2         2       1.00        0.500000    0.000017
3         4       3.00        0.750000    0.000044
4         3       2.00        0.666667    0.000037
5         4       4.50        1.125000    0.000121
6         3       2.25        0.750000    0.000038
7         3       1.50        0.500000    0.000024
8         3       1.00        0.333333    0.000040
9         3       2.50        0.833333    0.000042


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **2. ALGEBRA (EQUATIONS)**

</div>

In [7]:
# our price model - price depends on sqft and bedrooms, plus a base amount
# Price = a*(sqft) + b*(bedrooms) + c
equation_str = "Price = a*sqft_living + b*bedrooms + c"
print(equation_str)

# a, b, c are unknowns - we don't know their values yet
# a = how much price increases per extra sqft
# b = how much price increases per extra bedroom
# c = base price (like a starting price even with 0 sqft, just theoretically)
print("a = price per sqft")
print("b = price per bedroom")
print("c = base price / intercept")

Price = a*sqft_living + b*bedrooms + c
a = price per sqft
b = price per bedroom
c = base price / intercept


In [8]:
# building X matrix (features) and y vector (target price)
X = df[['sqft_living', 'bedrooms']].to_numpy(dtype=float)
X = np.column_stack([X, np.ones(len(X))])   # adding a column of 1s for c
y = df['price'].to_numpy(dtype=float)

# solving for a, b, c using least squares (best fit across ALL rows, not just 3)
coeffs, residuals, rank, sv = np.linalg.lstsq(X, y, rcond=None)
a, b, c = coeffs

print(f"a (price per sqft):     {a:.2f}")
print(f"b (price per bedroom):  {b:.2f}")
print(f"c (base price):         {c:.2f}")

print(f"\nFinal equation: Price = {a:.2f}*sqft_living + {b:.2f}*bedrooms + {c:.2f}")

a (price per sqft):     313.95
b (price per bedroom):  -57066.76
c (base price):         79469.36

Final equation: Price = 313.95*sqft_living + -57066.76*bedrooms + 79469.36


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **3. CREATE VECTORS & MATRICES (NUMPY)**

</div>

In [9]:
# features -> X matrix, add bias column for c
X = df[['sqft_living', 'bedrooms']].to_numpy(dtype=float)
X = np.column_stack([X, np.ones(len(X))])

# target -> y vector
y = df['price'].to_numpy(dtype=float)

print("X shape:", X.shape)
print("y shape:", y.shape)
print(X[:5])
print(y[:5])

X shape: (21613, 3)
y shape: (21613,)
[[1.18e+03 3.00e+00 1.00e+00]
 [2.57e+03 3.00e+00 1.00e+00]
 [7.70e+02 2.00e+00 1.00e+00]
 [1.96e+03 4.00e+00 1.00e+00]
 [1.68e+03 3.00e+00 1.00e+00]]
[221900. 538000. 180000. 604000. 510000.]


**Create X (features matrix) and y (target vector)** 

In [10]:
X = df[['sqft_living', 'bedrooms']].to_numpy(dtype=float)
X = np.column_stack([X, np.ones(len(X))])   # bias column for c

y = df['price'].to_numpy(dtype=float)

print("X shape:", X.shape)
print("y shape:", y.shape)
print(X[:5])
print(y[:5])

X shape: (21613, 3)
y shape: (21613,)
[[1.18e+03 3.00e+00 1.00e+00]
 [2.57e+03 3.00e+00 1.00e+00]
 [7.70e+02 2.00e+00 1.00e+00]
 [1.96e+03 4.00e+00 1.00e+00]
 [1.68e+03 3.00e+00 1.00e+00]]
[221900. 538000. 180000. 604000. 510000.]


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **4. MATRIX OPERATIONS**

</div>

In [11]:
A = df[['sqft_living', 'bedrooms', 'bathrooms']].to_numpy(dtype=float)[:3]
B = df[['sqft_living', 'bedrooms', 'bathrooms']].to_numpy(dtype=float)[3:6]

print("Matrix A:\n", A)
print("Matrix B:\n", B)

# Addition
print("\nA + B =\n", A + B)

# Subtraction
print("\nA - B =\n", A - B)

# Multiplication (element-wise)
print("\nA * B =\n", A * B)

# Multiplication (matrix/dot product)
print("\nA @ B.T =\n", A @ B.T)

Matrix A:
 [[1.18e+03 3.00e+00 1.00e+00]
 [2.57e+03 3.00e+00 2.25e+00]
 [7.70e+02 2.00e+00 1.00e+00]]
Matrix B:
 [[1.96e+03 4.00e+00 3.00e+00]
 [1.68e+03 3.00e+00 2.00e+00]
 [5.42e+03 4.00e+00 4.50e+00]]

A + B =
 [[3.14e+03 7.00e+00 4.00e+00]
 [4.25e+03 6.00e+00 4.25e+00]
 [6.19e+03 6.00e+00 5.50e+00]]

A - B =
 [[-7.80e+02 -1.00e+00 -2.00e+00]
 [ 8.90e+02  0.00e+00  2.50e-01]
 [-4.65e+03 -2.00e+00 -3.50e+00]]

A * B =
 [[2.3128e+06 1.2000e+01 3.0000e+00]
 [4.3176e+06 9.0000e+00 4.5000e+00]
 [4.1734e+06 8.0000e+00 4.5000e+00]]

A @ B.T =
 [[ 2312815.     1982411.     6395616.5  ]
 [ 5037218.75   4317613.5   13929422.125]
 [ 1509211.     1293608.     4173412.5  ]]


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **5. SOLVE SYSTEM OF EQUATIONS**

</div>

In [12]:
# taking first 3 houses from dataset
sample = df.iloc[:3]

# building matrix A (sqft, bedrooms, and 1)
A = np.array([
    [sample['sqft_living'].iloc[0], sample['bedrooms'].iloc[0], 1],
    [sample['sqft_living'].iloc[1], sample['bedrooms'].iloc[1], 1],
    [sample['sqft_living'].iloc[2], sample['bedrooms'].iloc[2], 1]
])

# putting prices into B
B = np.array([
    sample['price'].iloc[0],
    sample['price'].iloc[1],
    sample['price'].iloc[2]
])

print("A:\n", A)
print("B:", B)

# METHOD 1: Inverse method
A_inverse = np.linalg.inv(A)
answer_1 = A_inverse @ B
print("\nMethod 1 (Inverse) - a, b, c:", answer_1)

# METHOD 2: NumPy's solve function
answer_2 = np.linalg.solve(A, B)
print("Method 2 (NumPy solve) - a, b, c:", answer_2)

A:
 [[1180    3    1]
 [2570    3    1]
 [ 770    2    1]]
B: [221900. 538000. 180000.]

Method 1 (Inverse) - a, b, c: [   227.41007194 -51338.1294964  107570.50359712]
Method 2 (NumPy solve) - a, b, c: [   227.41007194 -51338.1294964  107570.50359712]


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **6. ADVANCED LINEAR ALGEBRA**

</div>

In [13]:
# taking 3 houses, 3 features each (sqft, bedrooms, bathrooms)
M = df[['sqft_living', 'bedrooms', 'bathrooms']].to_numpy(dtype=float)[:3]

print("Matrix M:\n", M)

# DOT PRODUCT - combining two rows into a single number
dot_result = np.dot(M[0], M[1])
print("\nDot product of row 0 and row 1:", dot_result)

# MATRIX MULTIPLICATION
matmul_result = M @ M.T
print("\nMatrix multiplication (M @ M.T):\n", matmul_result)

# DETERMINANT - a single number that tells us if the matrix can be inverted
det_result = np.linalg.det(M)
print("\nDeterminant of M:", det_result)

# INVERSE - only works if determinant is not zero
if abs(det_result) > 1e-6:
    inverse_result = np.linalg.inv(M)
    print("\nInverse of M:\n", inverse_result)
else:
    print("\nMatrix can't be inverted (determinant is 0)")

Matrix M:
 [[1.18e+03 3.00e+00 1.00e+00]
 [2.57e+03 3.00e+00 2.25e+00]
 [7.70e+02 2.00e+00 1.00e+00]]

Dot product of row 0 and row 1: 3032611.25

Matrix multiplication (M @ M.T):
 [[1392410.     3032611.25    908607.    ]
 [3032611.25   6604914.0625 1978908.25  ]
 [ 908607.     1978908.25    592905.    ]]

Determinant of M: -1452.5000000000007

Inverse of M:
 [[ 1.03270224e-03  6.88468158e-04 -2.58175559e-03]
 [ 5.76592083e-01 -2.82271945e-01  5.85197935e-02]
 [-1.94836489e+00  3.44234079e-02  2.87091222e+00]]


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **7. EIGENVALUES & EIGENVECTORS**


In [14]:
# taking 3 features from dataset
features = df[['sqft_living', 'bedrooms', 'bathrooms']].to_numpy(dtype=float)

# covariance matrix - shows how features vary together
cov_matrix = np.cov(features, rowvar=False)
print("Covariance matrix:\n", cov_matrix)

# calculating eigenvalues and eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(cov_matrix)

print("\nEigenvalues:", eigenvalues)
print("\nEigenvectors:\n", eigenvectors)

# how much each eigenvalue contributes (as a percentage)
explained = eigenvalues / eigenvalues.sum()
print("\nVariance explained by each:", explained)

Covariance matrix:
 [[8.43533681e+05 4.92596040e+02 5.33811988e+02]
 [4.92596040e+02 8.65015010e-01 3.69527118e-01]
 [5.33811988e+02 3.69527118e-01 5.93151289e-01]]

Eigenvalues: [8.43534307e+05+0.j 5.87414724e-01+0.j 2.45279947e-01+0.j]

Eigenvectors:
 [[-9.99999629e-01+0.j -6.83831639e-04+0.j -5.23320126e-04+0.j]
 [-5.83967471e-04+0.j  9.85188473e-01+0.j -1.71473996e-01+0.j]
 [-6.32828300e-04+0.j  1.71473627e-01+0.j  9.85188507e-01+0.j]]

Variance explained by each: [9.99999013e-01+0.j 6.96372559e-07+0.j 2.90776205e-07+0.j]


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **8. SVD (SINGULAR VALUE DECOMPOSITION)**

</div>

In [15]:
# taking 3 features from dataset
features = df[['sqft_living', 'bedrooms', 'bathrooms']].to_numpy(dtype=float)

# applying SVD
U, S, Vt = np.linalg.svd(features, full_matrices=False)

print("U shape:", U.shape)
print("Singular values (S):", S)
print("Vt shape:", Vt.shape)

# checking we can rebuild the original matrix from U, S, Vt
rebuilt = U @ np.diag(S) @ Vt
print("\nRebuilt matrix matches original:", np.allclose(features, rebuilt))

# how much of the data's pattern is captured by the top value
top_value_importance = (S[0]**2) / np.sum(S**2)
print(f"\nTop singular value captures {top_value_importance:.1%} of the data's pattern")

U shape: (21613, 3)
Singular values (S): [3.34257768e+05 1.76536063e+02 7.40018872e+01]
Vt shape: (3, 3)

Rebuilt matrix matches original: True

Top singular value captures 100.0% of the data's pattern


<style>
.summary{
  margin:56px 0;
  font-family:sans-serif;
  background:#F7F5F0;
  padding:24px;
  border-radius:8px;
}
.summary h2{
  font-family:serif;
  font-size:22px;
  margin:0 0 4px;
  color:#A6472B;
  font-weight:700;
}
.summary .section-note{color:#5B6672; font-size:14px; margin:0 0 24px;}

.sgrid{
  display:grid; grid-template-columns:repeat(2,1fr); gap:14px; margin:20px 0 28px;
}
.scard{
  border:1px solid #D7DCE1; border-radius:6px; padding:16px 18px;
  background:#FFFFFF;
  opacity:0; transform:translateY(10px);
  animation:reveal .5s ease-out forwards;
}
.scard h4{margin:0 0 6px; font-size:14.5px; font-weight:600; color:#A6472B;}
.scard p{margin:0; font-size:13.5px; color:#5B6672; line-height:1.5;}
.scard code{font-family:monospace; font-size:12.5px; color:#1F6F78; background:#F3E4DF; padding:1px 5px; border-radius:3px;}

.sgrid .scard:nth-child(1){animation-delay:.02s;}
.sgrid .scard:nth-child(2){animation-delay:.09s;}
.sgrid .scard:nth-child(3){animation-delay:.16s;}
.sgrid .scard:nth-child(4){animation-delay:.23s;}
.sgrid .scard:nth-child(5){animation-delay:.30s;}
.sgrid .scard:nth-child(6){animation-delay:.37s;}
.sgrid .scard:nth-child(7){animation-delay:.44s;}
.sgrid .scard:nth-child(8){animation-delay:.51s;}

@keyframes reveal{ to{ opacity:1; transform:translateY(0);} }

@media (prefers-reduced-motion: reduce){
  .scard{animation:none; opacity:1; transform:none;}
}

.insight{
  font-size:14px; color:#1B2430; background:#F3E4DF;
  border-left:3px solid #A6472B; padding:14px 16px; border-radius:2px; margin:0;
}
.insight strong{color:#A6472B;}

@media (max-width:560px){
  .sgrid{grid-template-columns:1fr;}
}
</style>

<div class="summary">
  <h2>Final summary of learnings</h2>
  <p class="section-note">What each step contributed to the bigger picture.</p>
  <div class="sgrid">
    <div class="scard"><h4>Basic math</h4><p>Percentages, ratios and proportions describe relationships fast — like price-per-sqft or what share of homes are large.</p></div>
    <div class="scard"><h4>Algebraic equations</h4><p>Writing <code>Price = a·sqft + b·bed + c</code> turns raw numbers into a real mathematical model.</p></div>
    <div class="scard"><h4>Vectors &amp; matrices</h4><p>How that model becomes code — <code>X</code> holds the features, <code>y</code> holds the target.</p></div>
    <div class="scard"><h4>Matrix operations</h4><p>Add, subtract, multiply — the building blocks used to combine and transform data.</p></div>
    <div class="scard"><h4>Solving AX = B</h4><p>Inverse or <code>np.linalg.solve</code> — literally how the best-fit coefficients are found.</p></div>
    <div class="scard"><h4>Determinants &amp; inverses</h4><p>Tell us whether a system even has a unique solution — determinant ≈ 0 means it's unstable.</p></div>
    <div class="scard"><h4>Eigenvalues &amp; eigenvectors</h4><p>Reveal the directions of greatest variance — the mathematical basis of PCA.</p></div>
    <div class="scard"><h4>SVD</h4><p>Generalizes this to any matrix shape — the backbone of compression and recommendation systems.</p></div>
  </div>
  <p class="insight">
    <strong>Real-data insight:</strong> the fit gives a <em>negative</em> coefficient for bedrooms.
    Not a bug — it's <strong>multicollinearity</strong>: once sqft is already in the equation, an
    extra bedroom in the same square footage usually means smaller rooms, which the model reads
    as a small negative effect.
  </p>
</div>